src/main/java/me/minidigger/hangar/controller/PagesController.java

```diff
@@ -1,45 +1,155 @@
+        String pageName = getPageName(page, subPage);
+        ModelAndView mav = new ModelAndView("projects/pages/view");
+        ProjectData projectData = projectService.getProjectData(author, slug);
+        ProjectPage projectPage = ProjectPage.of(pagesSerivce.getPage(projectData.getProject().getId(), pageName));
+        mav.addObject("p", projectData);
+        ScopedProjectData sp = new ScopedProjectData();
+        sp.setPermissions(Permission.IsProjectOwner.add(Permission.EditPage));
+        mav.addObject("sp", sp);
+        mav.addObject("page", projectPage);
+        mav.addObject("parentPage");
+        mav.addObject("editorOpen", false);
+        pagesSerivce.fillPages(mav, projectData.getProject().getId());
+        return fillModel(mav);
     }
 
     @Secured("ROLE_USER")
-    @PostMapping("/{author}/{slug}/pages/{page}/delete")
-    public Object delete(@PathVariable Object author, @PathVariable Object slug, @PathVariable Object page) {
-        return null; // TODO implement delete request controller
+    @PostMapping({"/{author}/{slug}/pages/{page}/delete", "/{author}/{slug}/pages/{page}/{subPage}/delete"})
+    public RedirectView delete(@PathVariable String author, @PathVariable String slug, @PathVariable String page, @PathVariable(required = false) String subPage) {
+        String pageName = page + (subPage != null && !subPage.isEmpty() ? "/" + subPage : "");
+        ProjectData projectData = projectService.getProjectData(author, slug);
+        if (projectData == null) {
+            throw new ResponseStatusException(HttpStatus.NOT_FOUND);
+        }
+        ProjectPagesTable projectPage = pagesSerivce.getPage(projectData.getProject().getId(), pageName);
+        if (projectPage == null) {
+            throw new ResponseStatusException(HttpStatus.NOT_FOUND);
+        }
+        if (!projectPage.getIsDeletable()) {
+            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Cannot delete this page");
+        }
+        projectPageDao.get().delete(projectPage);
+        return new RedirectView(routeHelper.getRouteUrl("projects.show", author, slug));
     }
 
     @Secured("ROLE_USER")
-    @PostMapping("/{author}/{slug}/pages/{page}/edit")
-    public Object save(@PathVariable Object author, @PathVariable Object slug, @PathVariable Object page) {
-        return null; // TODO implement save request controller
+    @PostMapping(value = {"/{author}/{slug}/pages/{page}/edit", "/{author}/{slug}/pages/{page}/{subPage}/edit"}, consumes = MediaType.APPLICATION_FORM_URLENCODED_VALUE)
+    public Object save(@PathVariable String author, @PathVariable String slug, @PathVariable String page, @PathVariable(required = false) String subPage, @RequestParam(value = "parent-id", required = false) String parentId, @RequestParam("content") String pageContent, @RequestParam("name") String newPageName) {
```

Comment:
Return type

Reply:
save() returns a ResponseEntity if its a new page, or a RedirectView if its editing an existing page. so this has to be left as Object.